src/context.rs

```diff
@@ -30,19 +31,83 @@ impl<'a> QueryMatchContext<'a> {
         file_contents: &'a [u8],
         rule: &'a ResolvedRule,
         reported_any_violations: &'a AtomicBool,
+        context: &'a Context,
     ) -> Self {
         Self {
             path,
             file_contents,
             rule,
             reported_any_violations,
+            context,
         }
     }
 
     pub fn report(&self, violation: Violation) {
         self.reported_any_violations.store(true, Ordering::Relaxed);
         print_violation(&violation, self);
     }
+
+    pub fn get_node_text(&self, node: Node) -> &str {
+        node.utf8_text(self.file_contents).unwrap()
+    }
+
+    pub fn maybe_get_single_matching_node_for_query<'query, 'enclosing_node>(
```

Comment:
I'm sort of probably being blurry with some of these wrt the tree-sitter query interface that's "match-oriented" vs "capture-oriented"